### PR TITLE
Report Git LFS progress

### LINKED ISSUE #2226: Show Git LFS progress
We'll need to show it in a few different places:

- [x] Cloning.
- [x] Committing.
- [x] Checking out
  - [x] Reverting
  - [x] Switching branches

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #2226 

- [x] Cloning
- [x] Pushing
- [x] Pulling
- [x] Switching branches
- [x] Reverting

### CODE REVIEW

**Comment by joshaber** [comment]
This is blocked as Git LFS doesn't seem to report progress consistently, which makes testing this hard. The LFS team is looking into it.

**Comment by joshaber** [comment]
The Git LFS bugs were fixed in https://github.com/git-lfs/git-lfs/pull/2465 and should be out in the next release 💥 🐛

**Comment by joshaber** [comment]
The next Git LFS release won't be until mid next week at the earliest. We'll have to play it by ear, but I think it's OK to ship 1.0 without this and get it into a patch release shortly after.

**Comment by joshaber** [comment]
I found a problem with how LFS is reporting the number of files being downloaded: https://github.com/git-lfs/git-lfs/issues/2599

---

I've also run into a new and exciting UX problem.

`lfs clone` downloads files in parallel for ✨  **_super performance_** ✨ But this means it's really hard to show the user a unified, reliable progress bar. File X might be 50% done while File Y is 1% done.

We could show progress in terms of how many files are downloaded (that'd be blocked on https://github.com/git-lfs/git-lfs/issues/2599), but that'll be a little misleading, e.g., if we've downloaded all the small files but have 2 huge files left.

@desktop/design Thoughts?

**Comment by j-f1** [comment]
> We could show progress in terms of how many files are downloaded (that'd be blocked on git-lfs/git-lfs#2599), but that'll be a little misleading, e.g., if we've downloaded all the small files but have 2 huge files left.

Maybe *data downloaded*/*total data to download*?

**Comment by joshaber** [comment]
>Maybe data downloaded/total data to download?

Unfortunately we don't know that up front. We only know the size of the data for the currently downloading files.

**Comment by joshaber** [comment]
@niik and I talked this over and decided that we could ship a version where we display the indeterminate progress bar and the raw progress output. This is less than ideal, but better than what we currently have which is nothing.

**Comment by joshaber** [comment]
This is finally ready 🙏 

giphy-2

**Comment by joshaber** [comment]
Lil poke ☝️

File: app/src/lib/progress/lfs.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      isNaN(totalFiles) ||
+      isNaN(downloadedBytes) ||
+      isNaN(totalBytes)
+    ) {
+      return cannotParseResult
+    }
+
+    const verb = this.directionToHumanFacingVerb(direction)
+    const info: IGitProgressInfo = {
+      title: `${verb} "${name}" (${downloadedBytes} of ${totalBytes})…`,
```
[the remark this answers is not in the document]
Comment (joshaber) [reply app/src/lib/progress/lfs.ts]: That's intentional, see https://github.com/desktop/desktop/pull/2355#issuecomment-330556198.

File: app/src/ui/app.tsx [hunk 2]
```diff
@@ -1359,6 +1360,11 @@ export class App extends React.Component<IAppProps, IAppState> {
     }
 
     const state = selection.state
+    const revertProgress = state.revertProgress
+    if (revertProgress) {
+      return <RevertProgress progress={revertProgress} />
```
[the remark this answers is not in the document]
Comment (joshaber) [reply app/src/ui/app.tsx]: Yeah, I tested it by cloning https://github.com/git-lfs/git-lfs-test and then reverting the most recent commit a bunch of times.

File: app/src/ui/toolbar/index.tsx [hunk 3]
```diff
@@ -3,3 +3,4 @@ export * from './button'
 export * from './dropdown'
 export * from './push-pull-button'
 export * from './branch-dropdown'
+export { RevertProgress } from './revert-progress'
```
[the remark this answers is not in the document]
Comment (joshaber) [reply app/src/ui/toolbar/index.tsx]: I'm trying to be a bit pickier about what gets exported from `index`es, think of the `index` as exposing the public interface, but we might `export` other things from `revert-progress` for testing purposes.

That said, it's all just theoretical in this case. I can change it to `*` if you'd rather.

[the remark this answers is not in the document]
Comment (shiftkey) [reply app/src/ui/toolbar/index.tsx]: Nah, happy to leave it as-is.

**Comment by joshaber** [comment]
🍃

**Pull request merged**